java/vector/src/main/java/org/apache/arrow/vector/DecimalVector.java

```diff
@@ -206,41 +207,30 @@ public void set(int index, ArrowBuf buffer) {
   public void setBigEndian(int index, byte[] value) {
     BitVectorHelper.setValidityBitToOne(validityBuffer, index);
     final int length = value.length;
-    int startIndex = index * TYPE_WIDTH;
-    if (length == TYPE_WIDTH) {
-      for (int i = TYPE_WIDTH - 1; i >= 3; i -= 4) {
-        valueBuffer.setByte(startIndex, value[i]);
-        valueBuffer.setByte(startIndex + 1, value[i - 1]);
-        valueBuffer.setByte(startIndex + 2, value[i - 2]);
-        valueBuffer.setByte(startIndex + 3, value[i - 3]);
-        startIndex += 4;
-      }
 
-      return;
+    // do the bound check.
+    valueBuffer.checkBytes(index * TYPE_WIDTH, (index + 1) * TYPE_WIDTH);
+
+    long outAddress = valueBuffer.memoryAddress() + index * TYPE_WIDTH;
+    // swap bytes to convert BE to LE
+    for (int byteIdx = 0; byteIdx < length; ++byteIdx) {
+      PlatformDependent.putByte(outAddress + byteIdx, value[length - 1 - byteIdx]);
     }
 
-    if (length == 0) {
-      valueBuffer.setZero(startIndex, TYPE_WIDTH);
+    if (length == TYPE_WIDTH) {
       return;
     }
 
-    if (length < 16) {
-      for (int i = length - 1; i >= 0; i--) {
-        valueBuffer.setByte(startIndex, value[i]);
-        startIndex++;
-      }
-
+    if (length == 0) {
+      PlatformDependent.setMemory(outAddress, DecimalVector.TYPE_WIDTH, (byte)0);
```

Comment:
Currently, ArrowBuf#setZero is using the same approach.

Reply:
yes, i saw this. i'm not using the setZero to avoid an extra bounds check (already done once at the top of the function).